src/util/asmap.cpp

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <vector>
+#include <assert.h>
+#include <crypto/common.h>
+
+namespace {
+
+uint32_t DecodeBits(std::vector<bool>::const_iterator& bitpos, uint8_t minval, const std::vector<uint8_t> &bit_sizes)
+{
+    uint32_t val = minval;
+    bool bit;
+    for (std::vector<uint8_t>::const_iterator bit_sizes_it = bit_sizes.begin();
+        bit_sizes_it != bit_sizes.end(); ++bit_sizes_it) {
+        if (bit_sizes_it + 1 != bit_sizes.end()) {
+            bit = *bitpos;
+            bitpos++;
+        } else {
+            bit = 0;
+        }
+        if (bit) {
+            val += (1 << *bit_sizes_it);
+        } else {
+            for (int b = 0; b < *bit_sizes_it; b++) {
+                bit = *bitpos;
+                bitpos++;
+                val += bit << (*bit_sizes_it - 1 - b);
+            }
+            return val;
+        }
+    }
+    return -1;
```

Comment:
1. What happens when you return `-1` and the type is  `uint32_t`? Does it return `(2^32)-1`?

2. It doesn't seem like anything checks for `-1` to catch this error. 

3. If the call stack was `DecodeJump-->DecodeBits` and `DecodeBits` returns `-1` then the next operation would be `pos += jump;` where `jump=-1`. Does this then integer overflow `pos`?

Reply:
I'm thinking this should probably be ``assert(0)`` — the file is corrupted
@sipa ?

Reply:
Your opinion in other util related issues would be also useful.

Reply:
Feel free to tag me in other util related issues, happy to give a look and add my two cents.

Reply:
I do not think it is even possible to reach this. The
```c++
if (bit_sizes_it + 1 != bit_sizes.end()) {
    …
} else {
    bit = 0;
}
if (bit) {
    …
} else {
    …
    return val;
}
```
guarantees that the one-to-last bit size will always return a value.
(so yes, replacing with `assert(0)` would make sense)

Reply:
In addition to assert(0), we should provide console output saying something like "the asmap file is corrupted".

Reply:
Or refactor the function, moving the second for loop outside of the first, so that the unreachable code is no longer there:
```c++
uint32_t DecodeBits(std::vector<bool>::const_iterator& bitpos, uint8_t minval, const std::vector<uint8_t> &bit_sizes)                                                        
{
    uint32_t val = minval;
    bool bit;
    std::vector<uint8_t>::const_iterator bit_sizes_it = bit_sizes.begin();
    for (; (bit_sizes_it + 1) != bit_sizes.end(); ++bit_sizes_it) {
        bit = *bitpos;
        ++bitpos;
        if (bit) {
            val += (1 << *bit_sizes_it);
        } else {
            break;
        }
    }
    for (int b = 0; b < *bit_sizes_it; ++b) {
        bit = *bitpos;
        ++bitpos;
        val += bit << (*bit_sizes_it - 1 - b);
    }
    return val;
}
```